Fix #5455: Resolve crash in AudioViewModel by initializing state variables

TanishMoral11 opened the pull request:
**Explanation:**
Fixes #5455:
- This PR addresses issue #5455, which involves a crash caused by `kotlin.UninitializedPropertyAccessException` in `AudioViewModel` when the device is rotated during exploration.
- The fix ensures proper initialization of state variables in the `loadAudio()` function, which prevents the app from crashing due to uninitialized properties.

**Essential Checklist:**
- [x] The PR title and explanation each start with "Fix #5455: ".
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).


**Before Changes Video :** 
https://github.com/user-attachments/assets/649c9869-3ce6-4e0c-bafa-9769ff6d655f

**After Changes Video :** 
https://github.com/user-attachments/assets/6ded81c1-4846-4328-bd2e-b5c25420808b

**Screenshot of Passing Tests on Espresso:**
![Screenshot 2024-12-10 185043](https://github.com/user-attachments/assets/730e972b-65ab-4831-823a-eca0f0893376)

TanishMoral11 committed:
Fix crash in AudioViewModel by initializing state variables

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:
```diff
@@ -88,6 +90,8 @@ class AudioViewModel @Inject constructor(
    * @param allowAutoPlay If false, audio is guaranteed not to be autoPlayed.
    */
   private fun loadAudio(contentId: String?, allowAutoPlay: Boolean, reloadingMainContent: Boolean) {
+    // Check if 'state' is initialized before proceeding.
```
This comment is not needed as the code is self explanatory

TanishMoral11 replied:
Thanks @adhiamboperes , I'll remove that comment to maintain a clean codebase.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:
```diff
@@ -66,9 +66,11 @@ class AudioViewModel @Inject constructor(
-  fun setStateAndExplorationId(newState: State, id: String) {
-    state = newState
-    explorationId = id
+  fun setStateAndExplorationId(newState: State?, id: String?) {
+    if (newState != null && id != null) {
+      state = newState
+      explorationId = id
+    }
```
I have traced the usage of this function to: https://github.com/oppia/oppia-android/blob/f7e46519f54ef37871cefce27de1812bf4510493/app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt#L276-L283. Here, I can see that we perform a nullcheck for `id`, but not for `state`. As state is the variable throwing the uninitialized property exception, it might be worth finding a more indpeth reason why it is not being initialized on time. Or perharbs both values are not being initialized on time.

TanishMoral11 replied:
Hey @adhiamboperes, 
Thanks for the insight! I will also ensure to add null checks for both state and id to prevent the issue. 
I'll keep you updated on my progress!

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11!

I feel like this solution is not robust because it provides no explanation why the variable is not being initialized. It also does not provide an actual fallback. What should happen if the state and id are null?

TanishMoral11 committed:
Fix crash in AudioViewModel on rotation by ensuring state initialization

TanishMoral11 commented:
@adhiamboperes PTAL

TanishMoral11 commented:
Hi @adhiamboperes , could you take a look at this PR ?
Let me know if there's anything more I should adjust. 
Thanks!

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:
```diff
@@ -66,11 +67,16 @@ class AudioViewModel @Inject constructor(
 
-  fun setStateAndExplorationId(newState: State, id: String) {
+  fun setStateAndExplorationId(newState: State?, id: String?) {
+    if (newState == null || id == null) {
+      Log.e("AudioViewModel", "Failed to set state or id - parameters are null")
+      return
+    }
```
This is not needed. These values are passed in from the presenter, and we make an assumption that they are not null by the time they are passed in here.

Side note, we use our own logger, `oppiaLogger` for printing messages. It has w, i, and e just like `Log`.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:
```diff
@@ -66,11 +67,16 @@ class AudioViewModel @Inject constructor(
     explorationId = id
   }
 
+
```
Remove double newline.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:
```diff
@@ -88,6 +94,11 @@ class AudioViewModel @Inject constructor(
    * @param allowAutoPlay If false, audio is guaranteed not to be autoPlayed.
    */
   private fun loadAudio(contentId: String?, allowAutoPlay: Boolean, reloadingMainContent: Boolean) {
+    if (!::state.isInitialized || !::explorationId.isInitialized) {
+      Log.w("AudioViewModel", "Cannot load audio: state or explorationId is not initialized.")
+      return
+    }
```
Refactor this to:
```suggestion
     val targetContentId = when {
      !contentId.isNullOrEmpty() -> contentId
      ::state.isInitialized -> state.content.contentId
      else -> ""
    }
```

which is more robust, adds a check for `state.isInitialized` and provides a final fallback (`""`) if neither `contentId` nor `state.content.contentId` is available. This version helps avoid potential crashes if `state` or `contentId` is unavailable.

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11! PTAL at my suggestions. 

Do not resolve reviewer comments, so that the reviewer can verify that they have been addressed.

TanishMoral11 committed:
AudioViewModel: resolved changes

TanishMoral11 committed:
Remove Unnecessary Lines

TanishMoral11 commented:
Hey @adhiamboperes,

I've successfully addressed all the required changes as per the feedback. Could you kindly take a look at the updates when you get a chance?
Thanks!!

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/audio/AudioViewModel.kt:
```diff
@@ -1,5 +1,6 @@
 package org.oppia.android.app.player.audio
 
+import android.util.Log
```
I think this can be removed. Please confirm. Ktlint should have caught this.

TanishMoral11 commented:
Hey @adhiamboperes PTAL 
Thanks !!

TanishMoral11 commented:
Hey @adhiamboperes PTAL, 

Could you kindly review this PR . I’ve addressed all the issues, and there are no further changes required from my side. However, if you need any modifications, please let me know, and I’d be happy to make them.

Additionally, I have two other open PRs (#5563 and #5573) where I need to address some changes or modifications. Due to my upcoming college exams, I will be less active until December 6. I will be available to work on them after that.

Please let me know how I should proceed.
Because The Oppia bot automatically closes PRs if there is no activity, so I would like to understand how I can prevent my PR from being closed in such cases.

Thanks !!

adhiamboperes committed:
Merge branch 'develop' into fix-audio-crash-rotation

adhiamboperes commented:
@TanishMoral11, regularly updating your PR from develop should keep it open. You only need to do this at least once a week(using the update button at the bottom of the PR)

TanishMoral11 commented:
Hey @adhiamboperes PTAL

TanishMoral11 committed:
Merge branch 'develop' into fix-audio-crash-rotation

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11. Code changes like this always need to be tested to verify that they work.

The function under investigation is:
```kt
/**
   * Load audio based on the contentId.
   *
   * @param contentId If contentId is null, then state.content.contentId is used as default.
   * @param allowAutoPlay If false, audio is guaranteed not to be autoPlayed.
   */
  private fun loadAudio(contentId: String?, allowAutoPlay: Boolean, reloadingMainContent: Boolean) {
   val targetContentId = contentId ?: state.content.contentId
    val voiceoverMapping =
      state.recordedVoiceoversMap[targetContentId] ?: VoiceoverMapping.getDefaultInstance()

 ....
  }
```
Given that the error stack is:
```kt
2024-07-01 12:33:04.441 20689-20689/org.oppia.android E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.oppia.android, PID: 20689
    kotlin.UninitializedPropertyAccessException: lateinit property state has not been initialized
```

We know that state is not being initialized before use. So we made this change, per current PR:

```kt
    val targetContentId = when {
      !contentId.isNullOrEmpty() -> contentId
      ::state.isInitialized -> state.content.contentId
      else -> ""
    }
```

In order to test if this solution correctly fixes the issue, I wrote a test aimed at verifying that audio will still be playable on orientation change. The test is as follows:
```kt
@Test
  fun testAudioFragment_playAudio_configurationChange_checkAudioPaused() {
    addMediaInfo()
    launch<AudioFragmentTestActivity>(
      createAudioFragmentTestIntent(
        internalProfileId
      )
    ).use {
      testCoroutineDispatchers.runCurrent()

      waitForTheView(allOf(withId(R.id.play_pause_audio_icon), WithNonZeroDimensionsMatcher()))
        .perform(click())
      testCoroutineDispatchers.runCurrent()
      
      onView(withId(R.id.audio_progress_seek_bar)).perform(setProgress(100))
     
      onView(isRoot()).perform(orientationLandscape())
      testCoroutineDispatchers.runCurrent()
      onView(withId(R.id.play_pause_audio_icon))
        .check(matches(withContentDescription(context.getString(R.string.audio_play_description))))
    }
  }
  ```

However, this test still fails with the exception: 
```
kotlin.UninitializedPropertyAccessException: lateinit property state has not been initialized
at org.oppia.android.app.player.audio.AudioViewModel.loadAudio(AudioViewModel.kt:99)
at org.oppia.android.app.player.audio.AudioViewModel.loadMainContentAudio(AudioViewModel.kt:76)
at org.oppia.android.app.player.audio.AudioFragmentPresenter.subscribeToAudioLanguageLiveData$lambda-3(AudioFragmentPresenter.kt:158)
```

When I checked my code at `AudioViewModel.kt:99`, I discovered that it is  ` val voiceoverMapping =
      state.recordedVoiceoversMap[targetContentId] ?: VoiceoverMapping.getDefaultInstance()`. The state object here has not been initialized.

Your task is to find out why state has not been initialized, and run the test above against the fix until it passes. The test may need some modification, please use your discretion.

I created a [gist](https://gist.github.com/adhiamboperes/058764b5afdc2a4d50a24fc9790cce36) that shows you where `waitForTheView` and `WithNonZeroDimensionsMatcher()` come from.

TanishMoral11 committed:
Correct setStateAndExplorationId

TanishMoral11 committed:
Merge branch 'fix-audio-crash-rotation' of https://github.com/TanishMoral11/oppia-android into fix-audio-crash-rotation

TanishMoral11 committed:
Added Test

TanishMoral11 committed:
Improving Formatting

TanishMoral11 committed:
Improve Formatting

TanishMoral11 commented:
Hey @adhiamboperes PTAL, 

So, The Issue Was Caused By State Not Being Initialized On Time Due To Missing Or Delayed Checks In The Lifecycle Flow. I Have Investigated And Addressed The Initialization Process, Ensuring Null Checks Are Implemented For Both State And Id.

Additionally, I Refined The Test To Validate The Fix And Confirm The Code Now Behaves As Expected.

TanishMoral11 committed:
Solved Klint Issue

TanishMoral11 commented:
Hey @adhiamboperes PTAL !!

adhiamboperes commented on .idea/codeStyles/Project.xml:
This file should not be part of PR changes.

TanishMoral11 replied:
Done.

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/player/audio/AudioFragmentTest.kt:
```diff
@@ -208,14 +211,95 @@ class AudioFragmentTest {
+    }
+  }
+
+  /** Returns a matcher that matches view based on non-zero width and height. */
```
These helper classes should be moved after the last test in this file, to maintain readability of the class.

TanishMoral11 replied:
Done.

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11. While your latest test cases seem to be passing on Robolectric, it fails when I test it on Espresso on my emulator. Can you post a screenshot showing that the tests pass on espresso? Could you also show a video demo that this bug does not recurr with the kates code changes?

TanishMoral11 committed:
Remove .idea/codeStyles/Project.xml from commit

TanishMoral11 committed:
Change Helper Class Position In AudioFragmentTest.kt

TanishMoral11 commented:
Hey @adhiamboperes PTAL !!

Thank you for your feedback.

Before Changes Video : 
https://github.com/user-attachments/assets/649c9869-3ce6-4e0c-bafa-9769ff6d655f

After Changes Video : 
https://github.com/user-attachments/assets/6ded81c1-4846-4328-bd2e-b5c25420808b

Screenshot of Passing Tests :
![Screenshot 2024-12-10 185043](https://github.com/user-attachments/assets/730e972b-65ab-4831-823a-eca0f0893376)

adhiamboperes commented on .idea/codeStyles/Project.xml:
This is a project file, but should not be modified(not sure why this happens on windows). You can fix this by pulling `git pull upstream develop`. If there are conflicts, choose upstream so that the file is restored to it's original state. This PR should only have 2 changed files, the  viewmodel and the test file.

TanishMoral11 replied:
Done.

adhiamboperes reviewed (approved):
Thanks @TanishMoral11, this LGTM.

Just one last comment. PTAL.

TanishMoral11 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into fix-audio-crash-rotation

TanishMoral11 committed:
Reversed changes on Project.xml

TanishMoral11 committed:
Fixed Formatting

TanishMoral11 commented:
Hey @adhiamboperes , Could You PTAL

adhiamboperes committed:
Merge branch 'develop' into fix-audio-crash-rotation
